### PR TITLE
feat(parser): Add appsync resolver event Zod schemas

### LINKED ISSUE #3283: Feature request: Add AppSync Resolver Event Schema to Parser
### Use case

I would like to use the Powertools parser library to parse the AppSync lambda resolver event.  (The fields are described here: https://docs.aws.amazon.com/appsync/latest/devguide/resolver-context-reference.html) It would be nice to be able to use an off the shelf Zod schema from Powertools rather than hand-rolling the schema myself as I am currently doing. Happy to work on this PR myself.

### Solution/User Experience

It would work like any of the other built-in schemas that Powertools currently supports: https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/parser/#built-in-schemas.

### Alternative solutions

User can write the Zod schema themselves.

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by svozza**
## Summary

This PR adds Zod schemas for AppSync lambda resolvers and AppSync lambda batch resolvers.

### Changes
- Adds the Zod schema for the two event types
- Exports the TS types derived from the schema
- Adds tests, both positive and negative
- Adds example payloads as a doc string and links to the relevant AWS docs

**Issue number:** 
Resolves #3283.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Review by dreamorosi — COMMENTED** [review]
Thank you so much for opening this PR @svozza!

I have done a first quick review and it's looking good, only left a couple minor comments.

On Monday I will do a more comprehensive review with a test build of the utility and get back to you.

File: packages/parser/package.json
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parser/package.json:1]: Additionally, it'd be great if you could add dedicated exports for the new AppSync schemas similar to the existing ones under both `exports` and `typeVersions`. This way customers can import these schemas without having to bring in all the others in their bundle.

[the remark this answers is not in the document]
Comment (svozza) [reply packages/parser/package.json:1]: Sure thing. Just FYI, the formatting changes were done by the linter.

File: packages/parser/src/schemas/appsync.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    })
+    .nullable(),
+  stash: z.record(z.any()),
+});
+
+/**
+ * A zod schema for a batch AppSync resolver event
+ *
+ * @example
+ * /*
```
[the remark this answers is not in the document]
Comment (svozza) [reply packages/parser/src/schemas/appsync.ts]: *fixed

File: packages/parser/package.json
[the remark this answers is not in the document]
Comment (svozza) [reply packages/parser/package.json:1]: Done!

File: packages/parser/src/schemas/appsync.ts [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+ * @see {@link https://docs.aws.amazon.com/appsync/latest/devguide/resolver-context-reference-js.html}
+ */
+
+const AppSyncResolverSchema = z.object({
+  arguments: z.record(z.any()),
+  identity: z.optional(AppSyncIdentity),
+  source: z.record(z.any()).nullable(),
+  request: z.object({
+    headers: z.record(z.string()),
+  }),
```
Comment (leandrodamascena) [inline packages/parser/src/schemas/appsync.ts:114]:
```suggestion
  request: z.object({
    headers: z.record(z.string()),
    domainName: z.string().nullable(),
  }),
```

When using custom domain name AppSync sends this additional field.

File: packages/parser/src/schemas/appsync.ts [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+const AppSyncIamIdentity = z.object({
+  accountId: z.string(),
+  cognitoIdentityPoolId: z.string().nullable(),
+  cognitoIdentityId: z.string().nullable(),
+  sourceIp: z.array(z.string()),
+  username: z.string(),
+  userArn: z.string(),
+  cognitoIdentityAuthType: z.string().nullable(),
+  cognitoIdentityAuthProvider: z.string().nullable(),
```
Comment (leandrodamascena) [inline packages/parser/src/schemas/appsync.ts:11]: These fields are only present when you assume a role through a federation with Cognito Identity pools and then, right? If yes, make sense both be nullable.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
I'm not looking at the TS part because you guys are more expert than me in that area, but looking at the events it looks pretty good to me. I just made an observation because AppSync sends `domainName` when invoking with a custom domain name and when not invoking it is None.

Nice work and attention to detail @svozza! 👏

File: packages/parser/src/schemas/appsync.ts (hunk 2 above)
Comment (svozza) [reply packages/parser/src/schemas/appsync.ts:114]: Ah, good catch!

File: packages/parser/src/schemas/appsync.ts (hunk 3 above)
Comment (svozza) [reply packages/parser/src/schemas/appsync.ts:11]: During my testing I made some GQL queries through the AppSync console using IAM authentication and printed out the `identity` field and all those fields were null.

Comment (leandrodamascena) [reply packages/parser/src/schemas/appsync.ts:11]: Thanks for confirming, this is what I expected and it makes sense make them `nullable`. When getting credentials from a Cognito Identity Pool, these values ​​should be
cognitoIdentityAuthType: authenticated/unauthenticated
cognitoIdentityAuthProvider: the name of the provider.

File: packages/parser/src/schemas/appsync.ts (hunk 2 above)
Comment (svozza) [reply packages/parser/src/schemas/appsync.ts]: Done!

**Review by dreamorosi — APPROVED** [review]
Thank you for the PR @svozza - appreciate it!

**Pull request merged**